backend/data_fetcher.py

```diff
@@ -0,0 +1,170 @@
+# Copyright 2020 Google LLC
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# =============================================================================
+
+"""A module for fetching from multiple-level preprocessing."""
+
+import utils
+
+from level_slices_reader import LevelSlices
+from metadata import Metadata
+
+
+class DataFetcher:
+    """Class for for fetching from multiple-level preprocessing."""
+
+    def __init__(self, file_path, root_dir, preprocess_bucket=None):
+
+        self._rawfile = file_path
+        self._preprocess_bucket = preprocess_bucket
+
+        original_file_name = utils.get_file_name(file_path)
+        self._preprocess_dir = '/'.join([root_dir, original_file_name])
+
+    def is_preprocessed(self):
+        """Returns if the raw data is preprocessed.
+
+        Returns:
+            A boolean indicating if the raw file is preprocessed.
+        """
+        metadata = Metadata(self._preprocess_dir,
+                            bucket=self._preprocess_bucket)
+        error = metadata.load()
```

Comment:
Looks here is loading the top level metadata file.
Is that possible the preprocessing only finished the raw data slices, but the rest of the level are not done yet?

Reply:
No, top level metadata is not saved until the preprocessing is complete.